### PR TITLE
Improve vision transformer performance.

### PR DESCRIPTION
**Pull request by GhostScreaming**
## Description
**Improve vision transfomer performance with following methods:**
1. Set FLAGS_enable_cublas_tensor_op_math and FLAGS_gemm_use_half_precision_compute_type flags. These two flags have remarkable impact on CuBLAS BatchGEMM computation performance.
2. Add fused attention blocks, which can break the bottleneck of vit performance.

## Performance and Accuracy
**ViT-Base-224 training performance improvement using N2C16**
Paddle version: develop, paddle.version.commit: 992b30bad03448c3b2d1128fccc399f4116ac385
1. Origin performance vs. Set flag performance
7106.14 imgs/sec vs. 9557.32 imgs/sec （**+34.49%**）
2. Origin performance vs.Fused attention performance
7106.14 imgs/sec vs. 7913.51 imgs/sec（**+11.36%**）
3. Origin performance vs.Set Flag + Fused attention performance
7106.14 imgs/sec vs. 10883.87 imgs/sec（**+53.16%**）

**ViT-Base-224 evaluation acc@1 curve in 300 epochs training:**
1. Origin evaluation curve vs. Set flag evaluation cure
eval_acc_set_flag
4. Origin evaluation curve vs. Fused attention evaluation cure
eval_acc_fused_attn

### CODE REVIEW

File: ppfleetx/models/vision_model/vit/vit.py [hunk 1]
```diff
@@ -47,6 +51,64 @@
...
+                 drop=0.,
+                 attn_drop=0.,
+                 drop_path=0.,
+                 act_layer=nn.GELU,
+                 norm_layer='nn.LayerNorm',
+                 epsilon=1e-5):
+        super().__init__()
+
+        assert qk_scale is None, "Fused attention doesn't support qk_scale."
+        if isinstance(drop_path, float):
```
Comment (sneaxiy) [inline ppfleetx/models/vision_model/vit/vit.py]: `isinstance(drop_path, (float, int))`

File: ppfleetx/models/vision_model/vit/vit.py [hunk 2]
```diff
@@ -47,6 +51,64 @@
...
+            assert drop_path == 0.0, "Fused attention doesn't support drop_path."
+        elif isinstance(drop_path, list):
```
Comment (sneaxiy) [inline ppfleetx/models/vision_model/vit/vit.py]: `isinstance(drop_path, (list, tuple))`

File: ppfleetx/models/vision_model/vit/vit.py [hunk 3]
```diff
@@ -140,19 +203,39 @@ def __init__(self,
...
-                qk_scale=qk_scale,
-                drop=drop_rate,
-                attn_drop=attn_drop_rate,
-                drop_path=dpr[i],
-                norm_layer=norm_layer,
-                epsilon=epsilon) for i in range(depth)
-        ])
+        self.use_fused_attn = use_fused_attn
+        if self.use_fused_attn:
+            logger.info(
```
Comment (sneaxiy) [inline ppfleetx/models/vision_model/vit/vit.py:213]: Do we need this log?

File: ppfleetx/models/vision_model/vit/vit.py [hunk 4]
```diff
@@ -140,19 +203,39 @@ def __init__(self,
...
+                "ViT use fused attention. Fused attention model checkpoint will be" \
+                " saved in normal attention format for inference checkpoint export," \
+                " and its optimizer checkpoint keeps the same.")
+            self.blocks = nn.LayerList([
```
Comment (sneaxiy) [inline ppfleetx/models/vision_model/vit/vit.py]: How about unify the code like this:
```python
block_fn = FusedBlock if self.use_fused_attn else Block

self.blocks = nn.LayerList([block_fn(...) for i in range(depth)]
```

File: ppfleetx/models/vision_model/vit/vit.py [hunk 5]
```diff
@@ -47,6 +51,64 @@
...
+            attn_dropout_rate=attn_drop,
+            normalize_before=True,
+            epsilon=epsilon)
+
+        mlp_hidden_dim = int(dim * mlp_ratio)
+        self.mlp = FusedFeedForward(
+            d_model=dim,
+            dim_feedforward=mlp_hidden_dim,
+            dropout_rate=drop,
+            activation="gelu",
```
Comment (sneaxiy) [inline ppfleetx/models/vision_model/vit/vit.py]: You support "relu" and "gelu" in L76-L78, but you fix "gelu" here. Why?

File: ppfleetx/models/vision_model/vit/vit.py [hunk 6]
```diff
@@ -205,6 +288,76 @@ def forward(self, x):
...
+                   include_sublayers=True,
+                   structured_name_prefix="",
+                   use_hook=True):
+        state_dict = super().state_dict(destination, include_sublayers,
+                                        structured_name_prefix, use_hook)
+        if self.use_fused_attn:
+            new_dict = []
+            poped_keys = []
+            for key, value in state_dict.items():
+                if r'qkv_' in key:
```
Comment (sneaxiy) [inline ppfleetx/models/vision_model/vit/vit.py]: Too many if-else branches. Try to write codes like:


```
replaced_dict = {
    "qkv_": "qkv.",
    "linear_": "proj.",
    ....
}

for key, value in state_dict.items():
    for rkey, rvalue in replaced_dict.items():
        if rkey in key:
            # write your codes here
```

Same below in `set_state_dict`.

File: ppfleetx/models/vision_model/vit/vit.py [hunk 7]
```diff
@@ -205,6 +288,76 @@ def forward(self, x):
         x = self.head(x)
         return x
 
+    def state_dict(self,
+                   destination=None,
+                   include_sublayers=True,
+                   structured_name_prefix="",
+                   use_hook=True):
+        state_dict = super().state_dict(destination, include_sublayers,
```
Comment (sneaxiy) [inline ppfleetx/models/vision_model/vit/vit.py:306]: The `qkv.weight` in ViTAttention is of the shape `[embed_dim, 3 * embed_dim]`, but `attn.qkv_weight` in FusedAttention is of the shape `[3, num_head, head_dim, embed_dim]`. There should have a transpose between `qkv.weight` and `attn.qkv_weight`.

File: ppfleetx/models/vision_model/vit/vit.py [hunk 8]
```diff
@@ -384,6 +539,8 @@ def ViT_huge_patch14_224(**kwargs):
                 depth=32,
                 num_heads=16,
                 mlp_ratio=4,
+                qkv_bias=True,
```
Comment (sneaxiy) [inline ppfleetx/models/vision_model/vit/vit.py:549]: Why change this?

File: ppfleetx/configs/vis/vit/ViT_base_patch16_384_ft_in1k_2n16c_dp_fp16o2.yaml [hunk 9]
```diff
@@ -11,7 +11,7 @@ Engine:
   accumulate_steps: 1
   logging_freq: 10
   mix_precision:
-    use_pure_fp16: False
+    use_pure_fp16: True
```
Comment (sneaxiy) [inline ppfleetx/configs/vis/vit/ViT_base_patch16_384_ft_in1k_2n16c_dp_fp16o2.yaml:14]: Why change this?

File: ppfleetx/models/vision_model/vit/vit.py [hunk 10]
```diff
@@ -396,6 +553,8 @@ def ViT_huge_patch14_384(**kwargs):
...
```
Comment (sneaxiy) [inline ppfleetx/models/vision_model/vit/vit.py:563]: Why change this?

**Review by sneaxiy — APPROVED** [review]

File: projects/vit/run_finetune.sh [hunk 11]
```diff
@@ -13,5 +13,9 @@
 # limitations under the License.
 
 export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
+
+export FLAGS_enable_cublas_tensor_op_math=True
```
Comment (GuoxiaWang) [inline projects/vit/run_finetune.sh]: 是否在统一在 yaml 中加一个 Flags 的 section ，统一进行配置和设置？

File: ppfleetx/utils/config.py [hunk 12]
```diff
@@ -112,6 +112,31 @@ def process_global_configs(config):
             )
 
     global_cfg = config['Global']
+
+    def set_default_flags(flags):
```
Comment (GuoxiaWang) [inline ppfleetx/utils/config.py]: 不需要这么麻烦的。
flags_config = config['Global'].pop('flags',  {})
paddle.set_flags(**flags_config)

**Review by GuoxiaWang — APPROVED** [review]
LGTM
